### DISCUSSION TITLE
网关在在线角色4500+左右，偶发网关队列mqlen 几十万上百万长度

### DISCUSSION THREAD

**Discussion by ghost** [cloudwu/skynet#1895] [Q&A]
如何解决，网关也用master 跟slave的方式可以吗，还是说要起新的网关监听新的端口才行

**Comment by terry8210** [comment]
前端延时是否大？

**Comment by cloudwu** [comment]
说明客户端上传数据的速度超过了网关的处理能力。

如果用的 https://github.com/cloudwu/skynet/blob/master/service/gate.lua ，其实它没有太多耗时工作。它主要是把数据流分割成包，然后把数据指针转发给 agent :  https://github.com/cloudwu/skynet/blob/master/service/gate.lua#L25

如果这个行为都对 cpu 造成压力的话，你可以把数据分包的工作分到每个 agent 上。那么你需要自己改写一个 gateserver ，把  https://github.com/cloudwu/skynet/blob/master/service/gate.lua#L78  这里的 openclient 调用放到 agent 里。让每个 agent 自己去处理 fd 上的数据。

另外，你可以增加流量控制，方法是使用 socket.pause https://github.com/cloudwu/skynet/blob/master/lualib/skynet/socket.lua#L262
在你发现某个 fd 上的流量太大，无法及时处理时调用它。方法是通过 socket.warning 注册一个 callback ，在 socket 消息太多未能即时处理时就会触发 https://github.com/cloudwu/skynet/blob/master/lualib/skynet/socket.lua#L479

skynet 带的 gate 只是一个简单的范例，而且写的时候还没有 socket.pause 以及 socket.warning ，你可以根据需要完善自己的版本。

**Reply by cloudwu** [reply]
我看了一下，目前 gateserver 如果发现流量太大，是会把 warning 发到 watchdog 上的 : https://github.com/cloudwu/skynet/blob/master/service/gate.lua#L68

skynet 在流量超标时，以及后续恢复正常(处理完 buffer 中数据)时，均会触发 warning 。

所以，你可以在 watchdog 里根据收到的 warning 信号，暂停以及恢复对应的 fd ，使用 `skynet.socketdriver.pause` 和 `skynet.socketdriver.start` 即可。

**Reply by ghost** [reply]
后来发现原因了，agent收到要发给客户端的包以后，也转发给网关一起下发了，其实是可以在agent发送给客户端的。
